docs/design-proposals/2023-11 TraceQL Metrics.md

```diff
@@ -0,0 +1,112 @@
+---
+Authors: Martin Disibio (@mdisibio)
+Created: 2023 November
+Last updated: 2023-11-13
+---
+
+# TraceQL Metrics
+
+## Summary
+
+This design document describes TraceQL language additions for extracting time-series metrics like error rates and latencies from traces.   
+
+This document is **not** meant to be a complete language specification. Rather it is meant to invite the community to comment on the direction and capabilities.
+
+### Table of Contents
+
+- [Intro](#intro)
+- [First Stage](#first-stage)
+  - [Interval](#interval)
+  - [Grouping](#grouping)
+  - [Filtering](#filtering)
+- [Additional Stages](#additional-stages)
+- [Arithmetic](#arithmetic)
+- [Notes](#notes)
+  - [Step Interval](#step-interval)
+
+## Intro
+
+Let's look at an example straight away.  This query plots the error rate by service over time:  
+`{status = error } | rate() by (resource.service.name)`
+
+A normal TraceQL query finds matching spans and returns data points like time, duration, service name, status code, etc.  A metrics query at the core does the same thing: _find matching spans_, but instead of seeing individual results, they are aggregated over time to create time-series.  A metrics query is built by extending a regular query with metrics pipeline stages.  The structure looks like this:
+
+`<span query...> | <metrics query...>`
+
+An important understanding is that _any_ valid TraceQL query can be turned into a metrics query. This finds the call rates to service B but only when initiated upstream by service A:  
+`{resource.service.name="A"} >> {resouce.service.name="B"} | rate()`
+
+
+## First Stage
+
+The first stage in a metrics query is responsible for turning spans into time series. This is a unique intersection between the two data types and has specialized aggregation functions.  We use the `..._over_time()` suffix to distinguish them from their counterparts.
+
+| function                                   | description                |
+---------------------------------------------| ---------------------------- 
+|`rate()`                                    | The rate of spans per second
```

Comment:
The field could be optional for those when specifying one you could rate/sum that field instead of the span count.

`{ resource.service.name = "foo"} | rate(span.attribute.order_count) by (resource.pod)`

Reply:
Interesting idea.  Do you mean it would inner sum the attribute and calculate rate?   

Example: `{ } | rate(span.read_bytes)` calculates read throughput?